### PR TITLE
Allow specifying num_samples to RandomSampler even when replacement=False

### LINKED ISSUE #38032: Allow specifying num_samples to RandomSampler when replacement=False
## 🚀 Feature

Allow RandomSampler to accept a value for num_samples when replacement=False.  

## Motivation

Currently RandomSampler explicitly prevents the user from specifying num_samples when replacement=False. I don't see any reason why this should be the case. It seems perfectly reasonable to sample only N unique batches every epoch from the possible pool of all batches. 

## Pitch

The only changes necessary would be to remove this check:

```python
        if self._num_samples is not None and not replacement:
            raise ValueError("With replacement=False, num_samples should not be specified, "
                             "since a random permute will be performed.")
```

And change the line: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/sampler.py#L107 

from ` return iter(torch.randperm(n).tolist())` to ` return iter(torch.randperm(n).tolist()[:self._num_samples])`

cc @SsnL

### PR DESCRIPTION
**Pull request by pm25**
Fixes #38032 #39214

Hi, I modified the RandomSampler to satisfy the requirement of #38032. I also added and deleted some test cases in the test/test_dataloader.py to match with the new requirement.

### CODE REVIEW

File: test/test_dataloader.py [hunk 1]
```diff
@@ -1456,18 +1456,22 @@ def sample_stat(sampler, num_samples):
...
-
-        self.assertRaises(ValueError, lambda: RandomSampler(self.dataset, num_samples=0))
+        # test sample without replacement and with specified num_samples
+        n = len(self.dataset) * 2
+        sampler_without_replacement = RandomSampler(self.dataset, num_samples=n)
+        count_repeated, minval, maxval, count_total = sample_stat(sampler_without_replacement, len(self.dataset))
+        self.assertTrue(count_repeated == len(self.dataset))
+        self.assertTrue(minval == 0)
+        self.assertTrue(maxval == len(self.dataset) - 1)
+        self.assertTrue(count_total == len(self.dataset) * 2)
```
Comment (ejguan) [inline test/test_dataloader.py:1474]: Could you please also add two tests here?
- `n < len(self.dataset)`
- `n` is not dividable by `len(self.dataset)` without remainder.

File: test/test_dataloader.py [hunk 2]
```diff
@@ -1500,6 +1504,33 @@ def test_random_sampler_len_with_replacement(self):
...
+        batch_size = 1
+        count_num_samples_in_data_loader = len(self._get_data_loader(
+            self.dataset, batch_size=batch_size, sampler=sampler))
+        self.assertEqual(num_samples, count_num_samples_in_data_loader)
+
+        # test with dataloader, batch_size = 6
+        batch_size = 6
+        count_num_samples_in_data_loader = len(self._get_data_loader(
+            self.dataset, batch_size=batch_size, sampler=sampler))
+        self.assertEqual(int(math.ceil(float(num_samples) / batch_size)),
```
Comment (ejguan) [inline test/test_dataloader.py:1531]: super nit: `num_samples // batch_size + (num_samples % batch_size > 0)` can get an accurate result here.

**Review by ejguan — APPROVED** [review]
Thanks for adding the patch. Looks good to me in general. Leaving two nit comments for testes.

**Comment by PM25** [comment]
ok, no problem.
I just added the test cases.

**Review by ejguan — APPROVED** [review]
Thank you! LGTM.

**Comment by PM25** [comment]
Thanks for the review!

**Pull request closed without merging**